src/mlpack/methods/decision_tree/random_binary_numeric_split_impl.hpp

```diff
@@ -125,8 +125,127 @@ double RandomBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
   if (gain < bestFoundGain && splitIfBetterGain)
     return DBL_MAX;
 
-  classProbabilities.set_size(1);
-  classProbabilities(0) = randomPivot;
+  splitInfo.set_size(1);
+  splitInfo[0] = randomPivot;
+
+  if (UseWeights)
+    gain /= totalWeight;
+  else
+    gain /= labels.n_elem;
+
+  return gain;
+}
+
+template<typename FitnessFunction>
+template<bool UseWeights, typename VecType, typename WeightVecType>
+double RandomBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const arma::Row<double>& labels,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    double& splitInfo,
+    AuxiliarySplitInfo& /* aux */,
+    const bool splitIfBetterGain)
+{
+  double bestFoundGain = std::min(bestGain + minimumGainSplit, 0.0);
+  // Forcing a minimum leaf size of 1 (empty children don't make sense).
+  const size_t minimum = std::max(minimumLeafSize, (size_t) 1);
+
+  // First sanity check: if we don't have enough points, we can't split.
+  if (data.n_elem < (minimum * 2))
+    return DBL_MAX;
+  if (bestGain == 0.0)
+    return DBL_MAX; // It can't be outperformed.
+
+  typename VecType::elem_type maxValue = arma::max(data);
+  typename VecType::elem_type minValue = arma::min(data);
+
+  // Sanity check: if the maximum element is the same as the minimum, we
+  // can't split in this dimension.
+  if (maxValue == minValue)
+    return DBL_MAX;
+
+  double totalWeight = 0.0;
+  double totalLeftWeight = 0.0;
+  double totalRightWeight = 0.0;
+  if (UseWeights)
+  {
+    totalWeight = arma::accu(weights);
+    bestFoundGain *= totalWeight;
+  }
+  else
+  {
+    bestFoundGain *= data.n_elem;
+  }
+
+  // Picking a random pivot to split the dimension.
+  double randomPivot = math::Random(minValue, maxValue);
```

Comment:
I am indeed following the algorithm described in the paper. It is mentioned on page 4 [here](https://orbi.uliege.be/bitstream/2268/9357/1/geurts-mlj-advance.pdf). But anyway, if there is such a big imbalance in data, then it makes perfect sense to split it about that point. This might be overfitting, but the split does make sense.

Reply:
Ahhh, ok, then let's stick with the paper's implementation then. :+1: